LoopKit/CarbKit/CarbStore.swift

```diff
@@ -289,78 +321,69 @@ public final class CarbStore: HealthKitSampleStore {
-                completion(self.getCachedCarbEntries().filterDateRange(start, end))
+    ///   - start: The earliest date of values to retrieve
+    ///   - end: The latest date of values to retrieve, if provided
+    /// - Returns: An array of carb entries, in chronological order by startDate, or error
+    private func getCarbEntries(start: Date? = nil, end: Date? = nil) -> CarbStoreResult<[StoredCarbEntry]> {
+        dispatchPrecondition(condition: .onQueue(queue))
+
+        var entries: [StoredCarbEntry] = []
+        var error: CarbStoreError?
+
+        cacheStore.managedObjectContext.performAndWait {
+            do {
+                entries = try self.getActiveCachedCarbObjects(start: start, end: end).map { StoredCarbEntry(managedObject: $0) }
+            } catch let coreDataError {
+                error = .coreDataError(coreDataError)
             }
-            return
         }
-        #endif
 
-        getCarbSamples(start: start, end: end) { (result) in
-            switch result {
-            case .success(let samples):
-                completion(samples)
-            case .failure:
-                self.queue.async {
-                    completion(self.getCachedCarbEntries().filterDateRange(start, end))
-                }
-            }
+        if let error = error {
+            return .failure(error)
         }
+
+        return .success(entries)
     }
 
-    /// Retrieves carb entries from HealthKit within the specified date range
+    /// Retrieves active (not superceded, non-delete operation) cached carb objects within the specified date range
     ///
     /// - Parameters:
     ///   - start: The earliest date of values to retrieve
     ///   - end: The latest date of values to retrieve, if provided
-    ///   - completion: A closure calld once the values have been retrieved
-    ///   - result: An array of carb entries, in chronological order by startDate
-    public func getCarbEntries(start: Date, end: Date? = nil, completion: @escaping (_ result: CarbStoreResult<[StoredCarbEntry]>) -> Void) {
-        getCarbSamples(start: start, end: end) { (result) in
-            switch result {
-            case .success(let samples):
-                completion(.success(samples))
-            case .failure(let error):
-                completion(.failure(error))
-            }
+    /// - Returns: An array of cached carb objects
+    private func getActiveCachedCarbObjects(start: Date? = nil, end: Date? = nil) throws -> [CachedCarbObject] {
+        dispatchPrecondition(condition: .onQueue(queue))
+
+        var predicates = [NSPredicate(format: "operation != %d", Operation.delete.rawValue),
+                          NSPredicate(format: "supercededDate == NIL")]
+        if let start = start {
+            predicates.append(NSPredicate(format: "startDate >= %@", start as NSDate))
+        }
+        if let end = end {
+            predicates.append(NSPredicate(format: "startDate < %@", end as NSDate))
```

Comment:
Most date range queries in Loop are inclusive on the earlier end and exclusive on the later end. However, it really doesn't much matter since (I believe) the underlying storage (sqlite) has millisecond precision. Adding in the `=` is really only going to affect events that happen exactly on the endDate down to the millisecond.